File: app/src/main/java/com/example/fabio/myproof/Bracket.java
package com.example.fabio.myproof;

import java.util.ArrayList;
import java.util.Arrays;

import static android.text.TextUtils.join;

/**
 * Created by fabio on 29/03/2017.
 */

class Bracket extends ArrayList<String> {

    Bracket() {}
    public void set(String source) {
        clear();
        if (source.isEmpty()) return;
        addAll(Arrays.asList(source.split(" ")));
    }
    public String toString() {return join(" ",this);}
    public boolean check(Command command) {
        boolean output = false;
        for (String item:this) {
            switch (item.replace("-","")) {
                case "Term":
                    item += "Variable";
                    break;
                case "Premise":
                    item += "Statement";
                    break;
                case "Token":
                    item += "VariableTermStatementPremiseSequent";
                    break;
            }
            if (item.startsWith("-")) {
                output &= !item.replace("-","").contains(command.output());
                output &= !item.replace("-","").equals(command.name);
            } else {
                output |= item.contains(command.output());
                output |= item.equals(command.name);
            }
        }
        return output;
    }
}

File: app/src/main/java/com/example/fabio/myproof/Bracket.kt
package com.example.fabio.myproof

import java.util.ArrayList
import java.util.Arrays

/**
 * Created by fabio on 29/03/2017.
 */

internal class Bracket : ArrayList<String>() {
    fun set(source: String) {
        clear()
        if (source.isEmpty()) return
        addAll(Arrays.asList(*source.split(" ".toRegex()).dropLastWhile { it.isEmpty() } .toTypedArray()))
    }

    override fun toString(): String = joinToString(" ")

    fun check(command: Command): Boolean {
        var output = false
        val iter = listIterator()
        while(iter.hasNext()) {
            val item = iter.next()
            when (item.replace("-", "")) {
                "Term" -> iter.set(item + "Variable")
                "Premise" -> iter.set(item + "Statement")
                "Token" -> iter.set(item + "VariableTermStatementPremiseSequent")
            }
            if (item.startsWith("-")) {
                output = output and !item.replace("-", "").contains(command.output())
                output = output and (item.replace("-", "") != command.name)
            } else {
                output = output or item.contains(command.output())
                output = output or (item == command.name)
            }
        }
        return output
    }
}
